### PR TITLE
feat(deps): pick and choose parts of brew bundle exec

### PR DESCRIPTION
**Pull request by mistydemeo**
This replaces the invocation to `brew bundle exec` with a lightweight Rusty replacement. Rather than running our entire build within a `brew bundle exec` subprocess, we effectively replace it using data that `brew bundle` gives us. We accept the following environment variables unconditionally:

* `PKG_CONFIG_PATH`
* `PKG_CONFIG_LIBDIR`
* `CMAKE_INCLUDE_PATH`
* `CMAKE_LIBRARY_PATH`

In addition, based on the dependencies contained within `HOMEBREW_DEPENDENCIES` that we receive from `brew bundle`, we calculate the following:

* Paths to add to `PATH`
* Library search paths to add to `RUSTFLAGS`

What we *don't* do, which `brew bundle exec` does, is inherit Homebrew's compiler shims. This means that we don't inherit a handful of things that Homebrew sets for us, but it also means that it doesn't capture references to the system compiler to try to inject extra flags we may decide we don't want.

I tested this with my `cue2ccd`; `cargo dist build --target=aarch64-apple-darwin --artifacts=local` works properly. Cross-compiling to aarch64 from x86_64 and vice versa should work if the Homebrew dependencies being requested are binaries to place in the `PATH`, which didn't work before.

This is a bit messy; I wrote it to get a working version out, but there's stuff to clean up and testing to do before it gets merged

Should address part of #498 - it's solution 2.
Closes #471.

### CODE REVIEW

File: cargo-dist/src/errors.rs [hunk 1]
```diff
@@ -259,6 +259,15 @@ pub enum DistError {
     /// random camino conversion error
     #[error(transparent)]
     FromPathBufError(#[from] camino::FromPathBufError),
+
+    /// Error parsing a string containing an environment variable
+    /// in VAR=value syntax
+    #[error("Unable to parse line as a key/value pair: {line}")]
+    #[diagnostic(help("key=value syntax expected"))]
+    EnvParseError {
```
Comment (Gankra) [inline cargo-dist/src/errors.rs:267]: The error message is a bit vague for an end-user

File: cargo-dist/src/lib.rs [hunk 2]
```diff
@@ -395,13 +396,133 @@ fn write_checksum(checksum: &str, src_path: &Utf8Path, dest_path: &Utf8Path) ->
...
+        match line.split_once('=') {
+            Some(parsed) => (key, value) = parsed,
+            None => {
+                return Err(DistError::EnvParseError {
+                    line: line.to_owned(),
+                })
+            }
+        };
+
+        parsed.insert(key, value);
```
Comment (Gankra) [inline cargo-dist/src/lib.rs:416]: a little simpler written as:

```suggestion
        let Some((key, value)) = line.split_once('=') else {
            return Err(DistError::EnvParseError {
                line: line.to_owned(),
            })
        };
        parsed.insert(key, value);
```

Comment (Gankra) [inline cargo-dist/src/lib.rs:416]: Is there established semantics for a key being defined multiple times? (This impl picks the last one)

File: cargo-dist/src/lib.rs [hunk 3]
```diff
@@ -395,13 +396,133 @@ fn write_checksum(checksum: &str, src_path: &Utf8Path, dest_path: &Utf8Path) ->
...
     );
 
+    let mut rustflags = target.rustflags.clone();
+    let mut desired_extra_env = vec![];
+    let skip_brewfile = env::var("DO_NOT_USE_BREWFILE").is_ok();
+    if let Some(brew) = &dist_graph.tools.brew {
+        if Utf8Path::new("Brewfile").exists() && !skip_brewfile {
+            // Uses `brew bundle exec` to just print its own environment,
+            // allowing us to capture what it generated and decide what
+            // to do with it.
```
Comment (Gankra) [inline cargo-dist/src/lib.rs:508]: Does this mean that the code to invoke `brew bundle exec` in `ci.yml.j2` can be removed? (If so, please do)

File: cargo-dist/src/lib.rs [hunk 4]
```diff
@@ -395,13 +396,133 @@ fn write_checksum(checksum: &str, src_path: &Utf8Path, dest_path: &Utf8Path) ->
...
+            let result = Command::new(&brew.cmd)
+                .arg("bundle")
+                .arg("exec")
+                .arg("--")
+                .arg("/usr/bin/env")
```
Comment (Gankra) [inline cargo-dist/src/lib.rs:513]: /usr/bin/env here is I assume a binary that just dumps all env-vars in scope when its ran, making this a simple way of telling bundle exec to tattle on itself?

File: cargo-dist/src/lib.rs [hunk 5]
```diff
@@ -395,13 +396,133 @@ fn write_checksum(checksum: &str, src_path: &Utf8Path, dest_path: &Utf8Path) ->
...
+                .output()
+                .into_diagnostic()
+                .wrap_err_with(|| "failed to exec brew bundle exec".to_string())?;
+
+            let env_output = String::from_utf8_lossy(&result.stdout).to_string();
```
Comment (Gankra) [inline cargo-dist/src/lib.rs:518]: Don't have a strong opinion, but curious if there is any particular reason why utf8_lossy is better/worse than "error out on non-utf8" in this situation?

File: cargo-dist/src/lib.rs [hunk 6]
```diff
@@ -438,6 +559,11 @@ fn build_cargo_target(dist_graph: &DistGraph, target: &CargoBuildStep) -> Result
             eprintln!(" --package={})", package);
         }
     }
+    // If we generated any extra environment variables to
+    // inject into the environment, apply them now.
+    for (key, value) in desired_extra_env {
+        command.env(key, value);
+    }
```
Comment (Gankra) [inline cargo-dist/src/lib.rs:566]: Checking you're aware of https://doc.rust-lang.org/std/process/struct.Command.html#method.envs ?

File: cargo-dist/src/lib.rs [hunk 7]
```diff
@@ -395,13 +396,133 @@ fn write_checksum(checksum: &str, src_path: &Utf8Path, dest_path: &Utf8Path) ->
...
+    if let Some(value) = environment.get("CMAKE_LIBRARY_PATH") {
+        desired_env.push(("CMAKE_LIBRARY_PATH".to_owned(), value.to_string()))
+    }
+    let mut paths = vec![];
+    // For each listed dependency, add it to the PATH
+    if let Some(optprefix) = environment.get("HOMEBREW_OPT") {
+        // Comma-separated list of all dependencies in the recursive tree
+        if let Some(dependencies) = environment.get("HOMEBREW_DEPENDENCIES") {
+            for dep in dependencies.split(',') {
+                let short_name = dep.split('/').next().unwrap();
```
Comment (Gankra) [inline cargo-dist/src/lib.rs:448]: Would be nice to not unwrap here, if possible

File: cargo-dist/src/lib.rs [hunk 8]
```diff
@@ -395,13 +396,133 @@ fn write_checksum(checksum: &str, src_path: &Utf8Path, dest_path: &Utf8Path) ->
...
+                paths.push(format!("{pkg_opt}/sbin"));
+            }
+        }
+    }
+    if !paths.is_empty() {
+        let our_path = env!("PATH");
+        let desired_path = format!("{our_path}:{}", paths.join(":"));
+
+        desired_env.insert(0, ("PATH".to_owned(), desired_path));
+    }
```
Comment (Gankra) [inline cargo-dist/src/lib.rs:463]: this part could use with a comment

(A++++ commenting in this PR, btw)

Comment (Gankra) [reply cargo-dist/src/lib.rs:463]: ohhh i see, maybe it doesn't need one

File: cargo-dist/src/lib.rs [hunk 9]
```diff
@@ -395,13 +396,133 @@ fn write_checksum(checksum: &str, src_path: &Utf8Path, dest_path: &Utf8Path) ->
...
+
+    desired_env
+}
+
+// Similar to the above, we read Homebrew's recursive dependency tree and
+// then append link flags to cargo-dist's rustflags.
+// These ensure that Rust can find C libraries that may exist within
+// each package's prefix.
```
Comment (Gankra) [inline cargo-dist/src/lib.rs:471]:
```suggestion
/// Similar to the above, we read Homebrew's recursive dependency tree and
/// then append link flags to cargo-dist's rustflags.
/// These ensure that Rust can find C libraries that may exist within
/// each package's prefix.
```

File: cargo-dist/src/lib.rs [hunk 10]
```diff
@@ -395,13 +396,133 @@ fn write_checksum(checksum: &str, src_path: &Utf8Path, dest_path: &Utf8Path) ->
...
+
+        parsed.insert(key, value);
+    }
+
+    Ok(parsed)
+}
+
+// Takes a BTreeMap of key/value environment variables produced by
+// `brew bundle exec` and decides which ones we want to keep for our own builds.
+// Returns a Vec containing (KEY, value) tuples.
```
Comment (Gankra) [inline cargo-dist/src/lib.rs:424]:
```suggestion
/// Takes a BTreeMap of key/value environment variables produced by
/// `brew bundle exec` and decides which ones we want to keep for our own builds.
/// Returns a Vec containing (KEY, value) tuples.
```

File: cargo-dist/src/lib.rs [hunk 11]
```diff
@@ -395,13 +396,133 @@ fn write_checksum(checksum: &str, src_path: &Utf8Path, dest_path: &Utf8Path) ->
...
+// These ensure that Rust can find C libraries that may exist within
+// each package's prefix.
+fn determine_brew_rustflags(base_rustflags: &str, environment: &BTreeMap<&str, &str>) -> String {
+    let mut rustflags = base_rustflags.to_owned();
+    // For each listed dependency, add it to CFLAGS/LDFLAGS
+    if let Some(optprefix) = environment.get("HOMEBREW_OPT") {
+        if let Some(dependencies) = environment.get("HOMEBREW_DEPENDENCIES") {
+            for dep in dependencies.split(',') {
+                let short_name = dep.split('/').next().unwrap();
+                let pkg_opt = format!("{optprefix}/{short_name}");
```
Comment (Gankra) [inline cargo-dist/src/lib.rs:479]: it's not a big deal but it does seem slightly odd that we parse this format twice (at very least could be factored to a reusable func)

**Review by Gankra — APPROVED** [review]
neato

File: cargo-dist/src/lib.rs (hunk 4 above)
Comment (mistydemeo) [reply cargo-dist/src/lib.rs:513]: Unfortunately, nope.

File: cargo-dist/src/lib.rs (hunk 3 above)
Comment (mistydemeo) [reply cargo-dist/src/lib.rs:508]: Yep! Thanks for the reminder, I actually missed adding that when I wrote this as quickly as I did.

File: cargo-dist/src/lib.rs (hunk 4 above)
Comment (Gankra) [reply cargo-dist/src/lib.rs:513]: Hmm, miscommunication? (wasn't asking if there was a better way, was just asking if i had understood how this was being used)

Comment (mistydemeo) [reply cargo-dist/src/lib.rs:513]: Oh sorry, yeah, I misread. `env` is a standard POSIX binary which, when called with no arguments, prints out the process's environment in `sh` syntax.

File: cargo-dist/src/lib.rs (hunk 7 above)
Comment (mistydemeo) [reply cargo-dist/src/lib.rs:448]: Unless I'm missing something, it's impossible for `split` on a string to not return at last a first element. Is there a more idiomatic way to say "get me the first element with no unwrap"? I feel like we don't need to manage an impossible `None` scenario.

Comment (Gankra) [reply cargo-dist/src/lib.rs:448]: great point! maybe just a comment reminding the reader that split is so nicely behaved, then

File: cargo-dist/src/lib.rs (hunk 2 above)
Comment (mistydemeo) [reply cargo-dist/src/lib.rs:416]: The string we're parsing from `env` will only ever return a given key once, so I'm trusting the input string to be well-formatted here. That said, I'll add a comment to that effect.

File: cargo-dist/src/lib.rs (hunk 11 above)
Comment (mistydemeo) [reply cargo-dist/src/lib.rs:479]: Yeah, I think I can DRY this up a bit.

**Pull request closed without merging**